File: appambit_sdk_flutter/android/src/main/kotlin/com/appambit/appambit_sdk_flutter/DatabaseFlutter.kt
package com.appambit.appambit_sdk_flutter

import android.content.Context
import com.appambit.sdk.AppAmbitDb
import com.appambit.sdk.models.db.DbResult
import com.appambit.sdk.models.db.DbStatement
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodChannel

class DatabaseFlutter {

    private lateinit var channel: MethodChannel

    fun attach(binding: FlutterPlugin.FlutterPluginBinding, @Suppress("UNUSED_PARAMETER") context: Context) {
        channel = MethodChannel(binding.binaryMessenger, "com.appambit/db")
        channel.setMethodCallHandler { call, result ->
            when (call.method) {
                "execute" -> {
                    val args = call.arguments as? Map<*, *>
                    val sql = args?.get("sql") as? String
                    if (sql == null) {
                        result.error("BAD_ARGS", "Missing 'sql'", null)
                        return@setMethodCallHandler
                    }
                    @Suppress("UNCHECKED_CAST")
                    val params = args["params"] as? List<Any?>

                    try {
                        val future = if (params != null && params.isNotEmpty()) {
                            AppAmbitDb.execute(sql, *params.toTypedArray())
                        } else {
                            AppAmbitDb.execute(sql)
                        }
                        future.then { dbResult ->
                            result.success(dbResultToMap(dbResult))
                        }
                        future.onError { e ->
                            result.success(errorMap(e.message ?: "Unknown error"))
                        }
                    } catch (e: Exception) {
                        result.success(errorMap(e.message ?: "Unknown error"))
                    }
                }

                "batch" -> {
                    val args = call.arguments as? Map<*, *>
                    @Suppress("UNCHECKED_CAST")
                    val statementsRaw = args?.get("statements") as? List<Map<String, Any?>>
                    val inTransaction = (args?.get("inTransaction") as? Boolean) ?: false

                    if (statementsRaw == null) {
                        result.error("BAD_ARGS", "Missing 'statements'", null)
                        return@setMethodCallHandler
                    }

                    val statements = statementsRaw.map { s ->
                        val stSql = s["sql"] as? String ?: ""
                        @Suppress("UNCHECKED_CAST")
                        val stParams = s["params"] as? List<Any?>
                        if (stParams != null && stParams.isNotEmpty()) {
                            DbStatement.of(stSql, *stParams.toTypedArray())
                        } else {
                            DbStatement.of(stSql)
                        }
                    }.toTypedArray()

                    try {
                        val future = if (inTransaction) {
                            AppAmbitDb.batchInTransaction(*statements)
                        } else {
                            AppAmbitDb.batch(*statements)
                        }
                        future.then { results ->
                            result.success(results.map { r -> dbResultToMap(r) })
                        }
                        future.onError { e ->
                            result.success(listOf(errorMap(e.message ?: "Unknown error")))
                        }
                    } catch (e: Exception) {
                        result.success(listOf(errorMap(e.message ?: "Unknown error")))
                    }
                }

                else -> result.notImplemented()
            }
        }
    }

    fun detach() {
        if (::channel.isInitialized) {
            channel.setMethodCallHandler(null)
        }
    }

    private fun dbResultToMap(r: DbResult): Map<String, Any?> = mapOf(
        "columns" to r.columns,
        "rows" to r.rows,
        "rowsRead" to r.rowsRead,
        "rowsWritten" to r.rowsWritten,
        "error" to if (r.hasError()) r.error else null,
    )

    private fun errorMap(msg: String): Map<String, Any?> = mapOf(
        "columns" to emptyList<String>(),
        "rows" to emptyList<List<Any?>>(),
        "rowsRead" to 0,
        "rowsWritten" to 0,
        "error" to msg,
    )
}

File: appambit_sdk_flutter/ios/Classes/DatabaseFlutter.swift
import Flutter
import AppAmbit

public class DatabaseFlutter {

    public static func execute(args: Any?, result: @escaping FlutterResult) {
        guard let argsDict = args as? [String: Any],
              let sql = argsDict["sql"] as? String, !sql.isEmpty else {
            result(FlutterError(code: "BAD_ARGS", message: "Missing 'sql'", details: nil))
            return
        }

        let params = argsDict["params"] as? [Any]

        let completion: (DbResult?, Error?) -> Void = { dbResult, error in
            DispatchQueue.main.async {
                if let error = error {
                    result(Self.errorMap(Self.errorMessage(error)))
                    return
                }
                guard let dbResult = dbResult else {
                    result(Self.errorMap("No result returned"))
                    return
                }
                result(Self.toMap(dbResult))
            }
        }

        if let params = params, !params.isEmpty {
            AppAmbitDb.execute(sql, params: params, completion: completion)
        } else {
            AppAmbitDb.execute(sql, completion: completion)
        }
    }

    public static func batch(args: Any?, result: @escaping FlutterResult) {
        guard let argsDict = args as? [String: Any],
              let statementsRaw = argsDict["statements"] as? [[String: Any]] else {
            result(FlutterError(code: "BAD_ARGS", message: "Missing 'statements'", details: nil))
            return
        }

        let inTransaction = (argsDict["inTransaction"] as? Bool) ?? false

        let statements = statementsRaw.map { s -> DbStatement in
            let sql = (s["sql"] as? String) ?? ""
            let params = s["params"] as? [Any]
            return DbStatement(sql: sql, params: params)
        }

        let completion: ([DbResult]?, Error?) -> Void = { results, error in
            DispatchQueue.main.async {
                if let error = error {
                    result([Self.errorMap(Self.errorMessage(error))])
                    return
                }
                result((results ?? []).map { Self.toMap($0) })
            }
        }

        if inTransaction {
            AppAmbitDb.batchInTransaction(statements, completion: completion)
        } else {
            AppAmbitDb.batch(statements, completion: completion)
        }
    }

    private static func toMap(_ r: DbResult) -> [String: Any] {
        let error: Any = r.error != nil ? r.error! : NSNull()
        return [
            "columns": r.columns,
            "rows": r.rows,
            "rowsRead": r.rowsRead,
            "rowsWritten": r.rowsWritten,
            "error": error,
        ]
    }

    private static func errorMessage(_ error: Error) -> String {
        let ns = error as NSError
        if ns.domain == "AppAmbit.ApiErrorType" {
            switch ns.code {
            case 1: return "Unauthorized"
            case 2: return "Network Unavailable"
            case 3: return "Unknown API error"
            default: return "API error (code \(ns.code))"
            }
        }
        return error.localizedDescription
    }

    private static func errorMap(_ msg: String) -> [String: Any] {
        return [
            "columns": [String](),
            "rows": [[Any]](),
            "rowsRead": 0,
            "rowsWritten": 0,
            "error": msg,
        ]
    }
}
